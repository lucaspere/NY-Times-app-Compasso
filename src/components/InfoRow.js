import React from 'react';
import { View, StyleSheet, Text } from 'react-native';
import dayjs from 'dayjs';
import localizedFormat from 'dayjs/plugin/localizedFormat';

dayjs.extend(localizedFormat);

export const InfoRow = ({ section, byline, publishedAt }) => {

   const formmatedDate = dayjs(publishedAt).format("llll");

   return (
      <View style={styles.containerFeed}>
         <View style={styles.authorContainer}>
            <Text numberOfLines={1} style={styles.byline}>{byline}</Text>
         </View>
         <View style={styles.articleInfo}>
            <Text style={styles.section}>{section}</Text>
            <Text style={styles.published_date}>{formmatedDate}</Text>
         </View>
      </View>
   )
}

const styles = StyleSheet.create({
   containerFeed: {
      height: 50,
      flexDirection: 'column',
      paddingHorizontal: 10,
      justifyContent: 'space-between',
      backgroundColor: '#051C2D',
   },
   section: {
      fontSize: 16,
      color: 'white'
   },
   articleInfo: {
      flexDirection: 'row',
      justifyContent: 'space-evenly',
      alignItems: 'center',
      padding: 5,
   },
   byline: {
      textAlign: 'center',
      fontSize: 16,
      color: 'white'
   },
   published_date: {
      fontSize: 16,
      color: 'white'
   },

});

export default InfoRow;