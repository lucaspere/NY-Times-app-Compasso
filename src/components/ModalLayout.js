import React from 'react';
import { Modal, TouchableOpacity, Text, View, StyleSheet } from 'react-native';
import { connect } from 'react-redux';
import { closeModal } from '../redux/actions/articleActions';
import { Linking } from 'expo';

export const ModalLayout = ({ title, abstract, link, closeModal, isOpen }) => {

   const TouchableOpacityStyle = {
      width: 80,
      height: 40,
      borderColor: '#fff',
      borderWidth: 2,
   }

   return (
      <Modal
         visible={isOpen}
         animationType="slide"
         onRequestClose={() => closeModal()}
      >
         <View style={styles.modalContainer}>
            <View style={styles.backBotton}>
               <TouchableOpacity
                  onPress={() => closeModal()}
                  style={TouchableOpacityStyle}
               >
                  <Text style={styles.textBotton}>Voltar</Text>
               </TouchableOpacity>
            </View>

            <View style={styles.bodyText}>
               <View style={{ marginTop: 20 }}>
                  <Text style={styles.modalTitle}>{title}</Text>
               </View>
               <View style={{ marginTop: 20 }}>
                  <Text
                     numberOfLines={1}
                     ellipsizeMode="tail"
                     style={styles.modalLink}
                     onPress={() => Linking.openURL(link)}>
                     {link}
                  </Text>
               </View>
               <View style={{ marginTop: 10 }}>
                  <Text style={styles.modalAbstract}>{abstract}</Text>
               </View>
            </View>
         </View >
      </Modal>
   )
}

const styles = StyleSheet.create({
   modalContainer: {
      flex: 1,
      backgroundColor: '#051C2D'
   },
   modalTitle: {
      fontSize: 30,
      color: "#051C2D",
      fontWeight: 'bold'
   },
   modalLink: {
      fontSize: 15,
      color: '#cc1235',
      textDecorationLine: 'underline',
   },
   modalAbstract: {
      fontSize: 18,
      color: "rgba(5,28,45,0.9)"
   },
   backBotton: {
      marginTop: 10,
      marginHorizontal: 15,
      borderBottomWidth: StyleSheet.hairlineWidth,
      borderBottomColor: 'yellow',
      height: 50
   },
   textBotton: {
      marginTop: 3.5,
      color: '#fff',
      fontSize: 18,
      fontWeight: 'bold',
      textAlign: 'center',
   },
   bodyText: {
      flex: 1,
      backgroundColor: "#fff",
      padding: 10,
   }
});

const mapStateToProps = ({ title, abstract, link, showModal }) => ({
   title,
   abstract,
   link,
   isOpen: showModal
});

const mapDispatchTopProps = (dispatch) => ({
   closeModal: () => dispatch(closeModal())
});

export default connect(mapStateToProps, mapDispatchTopProps)(ModalLayout)