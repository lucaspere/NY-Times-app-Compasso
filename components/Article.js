import React, { useState } from 'react';
import { View, ActivityIndicator, StyleSheet, ImageBackground, Text } from 'react-native';

import InfoRow from './InfoRow';

const Article = ({ title, image, section, subsection, byline }) => {

   const [loading, setLoading] = useState(true);

   return (
      <View style={styles.articleContainer}>
         <InfoRow
            section={section}
            subsection={subsection}
            byline={byline} />
         <ActivityIndicator
            style={StyleSheet.absoluteFill}
            animating={loading}
            size="large"
            color="#051C2D"
         />
         <ImageBackground
            source={image}
            style={styles.imageContainer}
            imageStyle={StyleSheet.absoluteFill}
            onLoad={() => setLoading(false)}
         >
            <View>
               <Text style={styles.title}>{title}</Text>
            </View>
         </ImageBackground>
      </View>
   )
}

const styles = StyleSheet.create({
   articleContainer: {
      marginVertical: 2,
      shadowColor: "#fff",
      shadowOffset: {
         width: 0,
         height: 8,
      },
      shadowOpacity: 0.44,
      shadowRadius: 10.32,
      elevation: 16,
   },
   imageContainer: {
      width: 'auto',
      height: 250,
      justifyContent: 'flex-end'
   },
   title: {
      color: "#fff",
      fontSize: 25,
      fontWeight: 'bold',
      textAlign: 'center',
      backgroundColor: 'rgba(0,0,0,0.35)'
   }
});


export default Article;