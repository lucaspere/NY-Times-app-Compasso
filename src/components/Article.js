import React, { useState } from 'react';
import { View, ActivityIndicator, StyleSheet, ImageBackground, Text, TouchableOpacity } from 'react-native';
import { connect } from 'react-redux';
import { articleStyle as styles} from '../styles/articleStyle';
import { setModal, setAbstract } from '../redux/actions/articleActions';

import InfoRow from './InfoRow';

const Article = (props) => {

   const [loading, setLoading] = useState(true);

   const { setAbstractValue, setModalValue } = props;

   const { title, image, section, byline, date, abstract } = props;

   return (
      <View style={styles.articleContainer}>
         <InfoRow
            publishedAt={date}
            section={section}
            byline={byline} />
         <ActivityIndicator
            style={StyleSheet.absoluteFill}
            animating={loading}
            size="large"
            color="#051C2D"
         />
         <ImageBackground
            source={image}
            style={styles.imageContainer}
            imageStyle={StyleSheet.absoluteFill}
            onLoad={() => setLoading(false)}
         >
            <TouchableOpacity
               onPress={() => {
                  setModalValue()
                  setAbstractValue(title, abstract);
               }}
            >
               <Text style={styles.title}>{title}</Text>
            </TouchableOpacity>
         </ImageBackground>
      </View>
   )
}

const mapStateToProps = ({ showModal }) => ({ showModal });
const mapDispatchTopProps = (dispatch) => ({
   setModalValue: () => dispatch(setModal()),
   setAbstractValue: (title, abstract) => dispatch(setAbstract(title, abstract))
})
export default connect(mapStateToProps, mapDispatchTopProps)(Article);