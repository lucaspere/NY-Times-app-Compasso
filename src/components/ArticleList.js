import React from 'react';
import { FlatList, View, ImageBackground, Text, StyleSheet } from 'react-native';

import {getKey} from '../utils/getKey';
import Article from './Article';

const ArticleList = ({ category, articles }) => {

   const keyExtractor = ({ uri }) => {

      const idPost = uri.match(/([\w|\d]+)-([\w|\d]+)-([\w|\d]+)-([\w|\d]+)-([\w|\d]+)/g);

      return idPost.toString();
   }
   const renderItem = ({ item}) => {

      if (item.section === category) {
         return (
            <Article
               abstract={item.abstract}
               date={item.published_date}
               title={item.title}
               section={item.section}
               byline={item.byline}
               image={
                  item.multimedia ? { uri: item.multimedia[0].url }
                     : require('../../assets/no-image-available.png')
               }
            />
         )
      }
   }

   return (
      <FlatList
         data={articles}
         renderItem={renderItem}
         keyExtractor={keyExtractor}
      />
   )
}



export default ArticleList