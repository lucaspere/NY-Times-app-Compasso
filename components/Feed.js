import React, { useState, useEffect } from 'react';
import {
   SafeAreaView,
   ActivityIndicator,
   Text,
   ViewPropTypes,
   StyleSheet,
   View,
   ImageBackground
} from 'react-native';

import ArticleList from './ArticleList';

import { data } from '../database/data'

const Feed = () => {

   const [isLoading, setIsLoading] = useState(false);
   const [error, setError] = useState(false);
   const [articles, setArticles] = useState([])


   const uri = data[1].multimedia[0].url
   if (isLoading) {
      return <ActivityIndicator styles={{ marginTop: 20 }} size="large" />
   }
   return (
      <SafeAreaView style={styles.containerFeed}>
         <ArticleList />
      </SafeAreaView>
   )
}

const styles = StyleSheet.create({
   containerFeed: {
      flex: 1,
      marginTop: 1,
      backgroundColor: "rgba(5,28,45,0.65)"
   },
});

export default Feed;