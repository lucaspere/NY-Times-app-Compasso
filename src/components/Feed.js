import React, { useState, useEffect } from 'react';
import {
   SafeAreaView,
   ActivityIndicator,
   StyleSheet,
   View,
   Text,
   TouchableOpacity
} from 'react-native';

import ArticleList from './ArticleList';

const Feed = ({ route, navigation, loading, error, articles}) => {

   if (loading) {
      return (
         <SafeAreaView style={styles.container}>
            <ActivityIndicator color='#fff' size="large" />
         </SafeAreaView>
      )
   }

   if (error.isError) {
      return (
         <SafeAreaView style={styles.errorContainer}>
            <View style={styles.errorMessageContainer}>
               <Text style={styles.errorMessage}>{error.errorMessage}</Text>
            </View>
            <TouchableOpacity
               style={styles.erroBotton}
               onPress={() => navigation.jumpTo(route.name, { name: "erroScreen" })}
            >
               <Text style={styles.buttonMessage}>Tentar novamente</Text>
            </TouchableOpacity>

         </SafeAreaView>
      )
   }

   return (
      <SafeAreaView style={styles.container}>
         <ArticleList articles={articles}
            category={route.name.toLowerCase()} />
      </SafeAreaView>
   )
}

const styles = StyleSheet.create({
   container: {
      flex: 1,
      marginTop: 1,
      backgroundColor: "rgba(5,28,45,0.65)"
   },
   errorContainer: {
      flex: 1,
      justifyContent: 'flex-start',
      alignItems: 'center',
      backgroundColor: '#EBE771'
   },
   errorMessageContainer: {
      backgroundColor: 'red',
      height: 40,
      marginTop: 40,
      margin: 15

   },
   errorMessage: {
      fontSize: 28,
      color: '#051C2D',
      fontWeight: 'bold'
   },
   erroBotton: {
      backgroundColor: 'white',
      width: 150,
      height: 50,
      alignItems: 'center',
      padding: 10,
      borderRadius: 1,
      borderBottomColor: 'red',
      borderBottomWidth: 4,
      borderRightColor: 'red',
      borderRightWidth: 2,
   },
   buttonMessage: {
      fontSize: 15,
      fontWeight: 'bold'
   }
});

export default Feed;