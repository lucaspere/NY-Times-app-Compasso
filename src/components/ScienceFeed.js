import React, { useState, useEffect } from 'react';
import {
   SafeAreaView,
   ActivityIndicator,
   StyleSheet,
} from 'react-native';
import { connect } from 'react-redux';

import { addSciencesArticles } from '../redux/actions/articleActions';
import ArticleList from './ArticleList';
import { fetchScienceArticles } from '../utils/api';

export const ScienceFeed = ({ route, addArticles, sciencesArticles }) => {

   const [loading, setLoading] = useState(true);

   useEffect(() => {
      const abortController = new AbortController();
      const signal = abortController.signal;

      fetchScienceArticles(signal, (err, articles) => {
         if (err) {
            return console.log(err.message)
         }
         addArticles(articles);
         setLoading(false)
      });
      return () => abortController.abort();
   }, [loading], [sciencesArticles]);

   if (loading) {
      return (
         <SafeAreaView style={styles.container}>
            <ActivityIndicator color='#fff' size="large" />
         </SafeAreaView>
      )
   }

   return (
      <SafeAreaView style={styles.container}>
         <ArticleList articles={sciencesArticles}
            category={route.name.toLowerCase()} />
      </SafeAreaView>
   )
}

const styles = StyleSheet.create({
   container: {
      flex: 1,
      marginTop: 1,
      backgroundColor: "rgba(5,28,45,0.65)"
   }
});

const mapStateToProps = ({ sciencesArticles }) => ({ sciencesArticles });
const mapDispatchToProps = (dispatch) => ({
   addArticles: (articles) => dispatch(addSciencesArticles(articles))
});

export default connect(mapStateToProps, mapDispatchToProps)(ScienceFeed);