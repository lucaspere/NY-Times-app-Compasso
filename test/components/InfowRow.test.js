import React from 'react';
import renderer from 'react-test-renderer';
import { InfoRow } from '../../src/components/InfoRow';

test('Deve renderizar corretanmente', () => {
   const tree = renderer.create(<InfoRow />).toJSON();

   expect(tree).toMatchSnapshot();
});


