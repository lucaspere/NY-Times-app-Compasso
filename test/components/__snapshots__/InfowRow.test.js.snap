// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deve renderizar corretanmente 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#051C2D",
      "flexDirection": "column",
      "height": 50,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
    }
  }
>
  <View>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "white",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "padding": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 16,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 16,
        }
      }
    >
      Sun, Mar 15, 2020 2:47 PM
    </Text>
  </View>
</View>
`;
